### DISCUSSION TITLE
question about ech SNI domain

### DISCUSSION THREAD

**Discussion by calvin2021y** [curl/curl#15406] [Q&A]
before ECH, there is a option to let your set the outer-domain.

with ECH, how can I set outer-domain ?

the example not from document say nothing about it.

```sh
curl --http3-only --curves X25519Kyber768Draft00 --ech true --doh-url https://1.1.1.1/dns-query https://cloudflare-ech.com/cdn-cgi/trace
```

**Comment by sftcd** [comment]
late here tonight, will look at the code tomorrow - the APIs I've proposed to OpenSSL do allow a command line override for the outer SNI (and I think that's a thing to preserve) but I forget details right now, sorry, back to you after some (sleepy:-) hours...

**Comment by calvin2021y** [comment]
I find this:

```md

"ecl:"
A base64 encoded ECHConfigList that is used for ECH.

"pn:"
A name to use to over-ride the public_name field of an ECHConfigList (only available with OpenSSL TLS support)
```

right now I am not able to build with openssl yet, will try fix it.

**Comment by sftcd** [comment]
Indeed, now I'm awake `--ech pn:` does the trick, e.g.

Without doing the override:
```
$ curl --ech true --doh-url https://one.one.one.one/dns-query https://min-ng.test.defo.ie:15443/echstat.php?format=json 
{"SSL_ECH_OUTER_SNI": "public.test.defo.ie","SSL_ECH_INNER_SNI": "min-ng.test.defo.ie","SSL_ECH_STATUS": "success", "date": "2024-10-25T10:47:37+00:00","config": "min-ng.test.defo.ie"}
```
And overriding the outer SNI:
```
$ curl --ech true --doh-url https://one.one.one.one/dns-query https://min-ng.test.defo.ie:15443/echstat.php?format=json --ech pn:xx.test 
{"SSL_ECH_OUTER_SNI": "xx.test","SSL_ECH_INNER_SNI": "min-ng.test.defo.ie","SSL_ECH_STATUS": "success", "date": "2024-10-25T10:46:22+00:00","config": "min-ng.test.defo.ie"}
```

Note that overriding the outer SNI can cause issues of course, e.g. cloudflare no longer like that at all, and a number of my test services at `*.test.defo.ie` use a haproxy instance on 443 to de-mux connections via the outer SNI so one gets odd behaviour if you play with it. (That's why the examples above aren't on port 443.)
